bank2/main.go

```diff
@@ -0,0 +1,267 @@
+				from = systemAccountID
+			} else {
+				to = systemAccountID
+			}
+		}
+		amount := rand.Intn(*maxTransfer)
+		start := time.Now()
+
+		if err := crdb.ExecuteTx(db, func(tx *sql.Tx) error {
+			startRead := time.Now()
+			rows, err := tx.Query(`SELECT id, balance FROM account WHERE id IN ($1, $2)`, from, to)
+			if err != nil {
+				return err
+			}
+			readDuration = time.Since(startRead)
+			for rows.Next() {
+				var id, balance int
+				if err = rows.Scan(&id, &balance); err != nil {
+					log.Fatal(err)
+				}
+				switch id {
+				case from:
+					fromBalance = balance
+				case to:
+					toBalance = balance
+				default:
+					panic(fmt.Sprintf("got unexpected account %d", id))
+				}
+			}
+			startWrite := time.Now()
+			if fromBalance < amount {
+				return nil
+			}
+			insert := `INSERT INTO transaction (id, txn_ref) VALUES ($1, $2);`
+			txnID := atomic.AddInt32(&txnCount, 1)
+			_, err = tx.Exec(insert, txnID, fmt.Sprintf("txn %d", txnID))
+			if err != nil {
+				return err
+			}
+			insert = `INSERT INTO transaction_leg (account_id, amount, running_balance, txn_id) VALUES ($1, $2, $3, $4);`
+			if _, err = tx.Exec(insert, from, -amount, fromBalance-amount, txnID); err != nil {
+				return err
+			}
+			if _, err = tx.Exec(insert, to, amount, toBalance+amount, txnID); err != nil {
+				return err
+			}
+			update := `UPDATE account SET balance = $1 WHERE id = $2;`
+			if _, err = tx.Exec(update, toBalance+amount, to); err != nil {
+				return err
+			}
+			if _, err = tx.Exec(update, fromBalance-amount, from); err != nil {
+				return err
+			}
+			writeDuration = time.Since(startWrite)
+			return nil
+		}); err != nil {
+			log.Printf("failed transaction: %v", err)
+			continue
+		}
+		if fromBalance >= amount {
+			readings <- measurement{read: readDuration, write: writeDuration, total: time.Since(start)}
+		}
+	}
+}
+
+func verifyTotalBalance(db *sql.DB) {
+	var sum int
+	if err := db.QueryRow("SELECT SUM(balance) FROM account").Scan(&sum); err != nil {
+		log.Fatal(err)
+	}
+	if sum != *numAccounts*initialBalance+initialSystemBalance {
+		log.Printf("The total balance is incorrect: %d.", sum)
+		os.Exit(1)
+	}
+}
+
+var usage = func() {
+	fmt.Fprintf(os.Stderr, "Usage of %s:\n", os.Args[0])
+	fmt.Fprintf(os.Stderr, "  %s <db URL>\n\n", os.Args[0])
+	flag.PrintDefaults()
+}
+
+func main() {
+	flag.Usage = usage
+	flag.Parse()
+
+	dbURL := "postgresql://root@localhost:26257/bank2?sslmode=disable"
+	if flag.NArg() == 1 {
+		dbURL = flag.Arg(0)
+	}
+
+	parsedURL, err := url.Parse(dbURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+	parsedURL.Path = "bank2"
+
+	db, err := sql.Open("postgres", parsedURL.String())
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() { _ = db.Close() }()
+
+	if _, err := db.Exec("CREATE DATABASE IF NOT EXISTS bank2"); err != nil {
+		log.Fatal(err)
+	}
+
+	// concurrency + 1, for this thread and the "concurrency" number of
+	// goroutines that move money
+	db.SetMaxOpenConns(*concurrency + 1)
```

Comment:
I think you also want to call `SetMaxIdleConns` to prevent connections from being recycled unnecessarily.